File: 1345.jump-game-iv.py
#
# @lc app=leetcode id=1345 lang=python3
#
# [1345] Jump Game IV
#

# @lc code=start
from collections import deque
from typing import List


class Solution:
    '''
    A graph problem. First build the graph, use index to identify nodes, link up the neighbour nodes. At the same time, record the values in a hash table. Then just BFS. There could be a lot of duplicates in the array. To avoid getting into an O(N^2) scenario, we keep track of what value has already been searched. I used an extra visited_values to ensure the loop through the index list will only occur once      
    '''
    def minJumps(self, arr: List[int]) -> int:
        n, hash_table = len(arr), {}
        for i, num in enumerate(arr):
            if num not in hash_table:
                hash_table[num] = []
            hash_table[num].append(i)
        
        dq, visited, visited_values, steps = deque(), [False]*n, set(), 0
        dq.append(0)
        visited[0] = True
        while dq:
            for i in range(len(dq)):
                node = dq.popleft()
                if node == n-1:
                    return steps                
                if arr[node] not in visited_values:
                    for nxt in hash_table[arr[node]]:
                        if not visited[nxt]:
                            dq.append(nxt)
                            visited[nxt] = True
                    visited_values.add(arr[node])
                if node > 0 and not visited[node-1]:
                    dq.append(node-1)
                    visited[node-1] = True
                if not visited[node+1]:
                    dq.append(node+1)
                    visited[node+1] = True
            steps += 1
        return -1
# @lc code=end

File: 1345.jump-game-iv.cs
/*
 * @lc app=leetcode id=1345 lang=csharp
 *
 * [1345] Jump Game IV
 */

// @lc code=start
using System.Collections.Generic;
public class Solution 
{
    public int MinJumps(int[] arr) 
    {
        int n = arr.Length;
        Dictionary<int, List<int>> hashTable = new Dictionary<int, List<int>>();
        for (int i = 0; i < n; i++)
        {
            if (!hashTable.ContainsKey(arr[i]))
            {
                hashTable.Add(arr[i], new List<int>());
            }
            hashTable[arr[i]].Add(i);
        }
        
        Queue<int> queue = new Queue<int>();
        bool[] visited = new bool[n];
        Array.Fill(visited, false);
        HashSet<int> visitedValues = new HashSet<int>();
        int steps = 0;
        queue.Enqueue(0);
        visited[0] = true;
        while (queue.Count > 0)
        {
            int m = queue.Count;
            for (int i = 0; i < m; i++)
            {
                int node = queue.Dequeue();
                if (node == n - 1)
                {
                    return steps;
                }
                if (!visitedValues.Contains(arr[node]))
                {
                    foreach (int nxt in hashTable[arr[node]])
                    {
                        if (!visited[nxt])
                        {
                            queue.Enqueue(nxt);
                            visited[nxt] = true;
                        }
                    }
                    visitedValues.Add(arr[node]);
                }
                if (node > 0 && !visited[node - 1])
                {
                    queue.Enqueue(node - 1);
                    visited[node - 1] = true;
                }
                if (!visited[node + 1])
                {
                    queue.Enqueue(node + 1);
                    visited[node + 1] = true;
                }
            }
            steps++;
        }
        return -1;
    }
}
// @lc code=end

